### PR TITLE
feat(git-fetcher): real npm-packlist semantics with .npmignore + bundleDependencies (#436 follow-up)

### PR DESCRIPTION
**Pull request by zkochan**
## Summary

Replaces the MVP `files`-field-only packlist with a faithful port of [`npm-packlist`](https://github.com/npm/npm-packlist) / [`pnpm/fs/packlist`](https://github.com/pnpm/pnpm/blob/94240bc046/fs/packlist/src/index.ts). Four passes:

1. **`.npmignore` + `.gitignore` walk** via `ignore::WalkBuilder` with per-directory inheritance. A `.npmignore` in `lib/` applies to `lib/**` only; the root `.gitignore` applies to the whole tree. `parents(false)` keeps ignore-file lookup confined to `pkg_dir` so a `.gitignore` *above* the package can't leak into the published file set.
2. **`files`-field allowlist** compiled into a single `Gitignore` matcher. Uses `matched_path_or_any_parents` so a pattern like `cli` matches both `bin/cli` (file match) and `lib/cli/index.js` (directory match, where `lib/cli` is a dir).
3. **Always-included files**: `package.json`, `README*`, `LICEN[SC]E*`, `CHANGES*`, `CHANGELOG*`, `HISTORY*`, `NOTICE*` at the root, plus `main` / `bin` paths. Override `.npmignore` and the `files`-field filter, matching npm-packlist's `alwaysIncluded`. `main` and `bin` are vetted through the always-excluded check before insertion — a manifest with `"main": "package-lock.json"` or `"bin": ".git/hook"` does not get to re-add cruft.
4. **`bundleDependencies` / `bundledDependencies` recursion** with cycle detection: each bundled dep gets its own packlist pass under `node_modules/<name>/`. Both spellings accepted plus the `bundleDependencies: true` shorthand. Cycle protection uses a `HashSet<PathBuf>` of canonical visited paths + a `MAX_BUNDLE_DEPTH = 32` belt-and-braces cap so a symlink loop or self-referencing bundle can't stack-overflow the fetcher. `bundleDependencies` names are validated against the same path-traversal discipline `cas_io::join_checked` uses (`Component::Normal` / `CurDir` only; rejects `..`, root, drive prefix).

### Always-excluded set, split by type

- `ALWAYS_EXCLUDED_DIR_SEGMENTS` — `.git`, `.svn`, `.hg`, `CVS`. Matched at any path segment, so VCS state is dropped at any depth.
- `ALWAYS_EXCLUDED_BASENAMES` — `.npmrc`, `npm-debug.log`, `.DS_Store`, `package-lock.json`, `yarn.lock`, `pnpm-lock.yaml`. Basename-only match, so a regular file like `lib/cvs-tools.txt` ships even though it mentions `CVS`.
- `ALWAYS_EXCLUDED_SUFFIXES` — `.orig` family.

### New runtime dep

`ignore = "0.4.25"` (BurntSushi, used by ripgrep; `Unlicense OR MIT`, `cargo deny` clean). Replaces the hand-rolled `*` / `**` / `?` matcher in the old `packlist.rs` — the new matcher uses `ignore::gitignore::GitignoreBuilder::add_line`, which gives full gitignore semantics (negation, anchored vs unanchored, directory-suffix rules) for free.

### Intentional upstream divergences (documented in the module header)

- When both `.npmignore` and `.gitignore` exist in the same directory, `ignore` combines their rules; npm-packlist would use only `.npmignore`. Same outcome for the common case (a `.npmignore` that's a strict superset); divergence only when `.npmignore` explicitly includes a path `.gitignore` excludes — rare in published packages.
- `.git/info/exclude` and global `~/.gitignore` are NOT honored. The fetcher operates on a clean tarball / checkout, not the user's working tree.

## Out of scope (still on #436)

- Two-slot CAS layout (`${filesIndexFile}\traw` + final) for skipping the re-import when packlist == raw. Pure perf optimization.
- §E real-npm test deferrals (need a `skip-if-no-npm` helper) + PATH-shim shallow-fetch test + Stage 2 resolver-side install tests.

## Test plan

- [x] `cargo nextest run -p pacquet-git-fetcher` — 21 packlist tests pass:
      - 7 carried over from the MVP packlist (files-field, *.orig, glob edge cases, etc.).
      - 7 new for `.npmignore` / `.gitignore` / `bundleDependencies` core behaviour: `npmignore_excludes_listed_paths`, `gitignore_excludes_when_no_npmignore`, `npmignore_does_not_drop_always_included_files`, `npmignore_in_subdir_applies_to_subtree_only`, `bundle_dependencies_subtree_is_included`, `bundled_dependencies_legacy_spelling_works`, `bundle_dependency_missing_dir_is_silently_skipped`.
      - 4 new safety regressions from review round 1: `npmignore_in_parent_dir_does_not_leak_in`, `bundle_dependencies_rejects_path_traversal`, `always_excluded_dir_segments_only_match_vcs`, `files_field_bare_basename_matches_at_depth`.
      - 3 new safety regressions from review round 2: `bundle_dependencies_self_cycle_is_caught` (symlink loop), `main_field_pointing_at_always_excluded_basename_is_refused`, `bin_field_pointing_at_vcs_segment_is_refused`.
- [x] `just ready` — 928 tests, 928 pass.
- [x] `RUSTDOCFLAGS=-D warnings cargo doc --no-deps --workspace --document-private-items` — clean.
- [x] `taplo format --check`
- [x] `just dylint`
- [x] `cargo deny check` — license + bans + advisories clean for the new `ignore` dep.

---
Written by an agent (Claude Code, claude-opus-4-7).

### CODE REVIEW

File: crates/git-fetcher/src/packlist.rs [hunk 1]
```diff
@@ -90,74 +94,236 @@ pub fn packlist(pkg_dir: &Path, manifest: &Value) -> Result<Vec<String>, Packlis
...
+    // already been deleted by [`crate::GitFetcher`] before this point.
+    let mut builder = WalkBuilder::new(pkg_dir);
+    builder
+        .standard_filters(false)
+        .hidden(false)
+        .git_ignore(true)
+        .git_exclude(false)
+        .git_global(false)
+        .require_git(false)
+        .add_custom_ignore_filename(".npmignore");
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/git-fetcher/src/packlist.rs:169]: Fixed in 8fd45daa. Added `.parents(false)` to the `WalkBuilder` config and a regression test (`npmignore_in_parent_dir_does_not_leak_in`) that plants a `.gitignore` in the temp parent and asserts `index.js` still ships. Good catch — the default would have meant a workspace-root gitignore could silently shrink the packlist.

File: crates/git-fetcher/src/packlist.rs [hunk 2]
```diff
@@ -90,74 +94,236 @@ pub fn packlist(pkg_dir: &Path, manifest: &Value) -> Result<Vec<String>, Packlis
...
+        }
+        let bundle_manifest = safe_read_package_json_from_dir(&bundle_pkg_dir)
+            .ok()
+            .flatten()
+            .unwrap_or_else(|| Value::Object(serde_json::Map::new()));
+        let nested = packlist(&bundle_pkg_dir, &bundle_manifest)?;
+        for rel in nested {
+            out.insert(format!("node_modules/{bundle_name}/{rel}"));
+        }
+    }
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/git-fetcher/src/packlist.rs:278]: Fixed in 8fd45daa. Added `is_safe_bundle_name` running the same `Path::components()` discipline as `cas_io::join_checked`: only `Normal` and `CurDir` pass, `ParentDir` / `RootDir` / `Prefix` are refused with a `tracing::warn!`. Scoped names like `@scope/foo` still pass since their `/` resolves *under* `node_modules`, not above it. Test: `bundle_dependencies_rejects_path_traversal` plants a sibling outside the package's node_modules and asserts `../escape` never makes it into the result.

File: crates/git-fetcher/src/packlist.rs [hunk 3]
```diff
@@ -90,74 +94,236 @@ pub fn packlist(pkg_dir: &Path, manifest: &Value) -> Result<Vec<String>, Packlis
...
+}
+
+fn should_always_exclude(rel: &str) -> bool {
     let basename = rel.rsplit('/').next().unwrap_or(rel);
     if ALWAYS_EXCLUDED_NAMES.contains(&basename) {
         return true;
     }
+    if rel.split('/').any(|seg| ALWAYS_EXCLUDED_NAMES.contains(&seg)) {
+        return true;
+    }
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/git-fetcher/src/packlist.rs]: Fixed in 8fd45daa. Split `ALWAYS_EXCLUDED_NAMES` into `ALWAYS_EXCLUDED_DIR_SEGMENTS` (`.git`, `.svn`, `.hg`, `CVS` — matched at any segment) and `ALWAYS_EXCLUDED_BASENAMES` (`.npmrc`, lockfiles, debug logs — basename only). The basename check on what was line 296 already covered file-typed entries everywhere, so the segment walk was dead duplicate work for them. Test: `always_excluded_dir_segments_only_match_vcs` plants both `lib/CVS/Root` (must exclude) and `lib/cvs-tools.txt` (must include) to pin the new behavior.

File: crates/git-fetcher/src/packlist.rs [hunk 4]
```diff
@@ -90,74 +94,236 @@ pub fn packlist(pkg_dir: &Path, manifest: &Value) -> Result<Vec<String>, Packlis
...
+/// pattern like `dist` matches both `dist` and everything under it.
+/// Bare basenames like `cli` should also match at any depth; the
+/// fallback below splits the candidate to try the leaf name.
+fn files_field_includes(matcher: &Gitignore, rel: &str) -> bool {
+    if matcher.matched(rel, false).is_ignore() {
         return true;
     }
-    ALWAYS_INCLUDED_PREFIXES.iter().any(|prefix| basename.starts_with(prefix))
+    rel.rsplit('/').next().is_some_and(|leaf| matcher.matched(leaf, false).is_ignore())
 }
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/git-fetcher/src/packlist.rs:342]: Fixed in 8fd45daa, and the writeup needs a correction: empirically (verified before the fix) `Gitignore::matched("bin/cli", false)` for pattern `cli` **does** return `Ignore` — unanchored gitignore patterns match at any depth. The leaf fallback was redundant.

BUT the test I added (`files_field_bare_basename_matches_at_depth`) caught a *different* bug: `Gitignore::matched` doesn't walk ancestor segments, so `files: ["cli"]` would NOT include `lib/cli/index.js` (where `lib/cli` is a directory the pattern matches). The fix swaps to `Gitignore::matched_path_or_any_parents` which walks the ancestor chain — that gives the directory-include semantics npm-packlist has. The new test pins root + nested-file + directory-match all three.

So the leaf fallback is gone, and the ancestor traversal is wired in correctly.

File: crates/git-fetcher/src/packlist.rs [hunk 5]
```diff
@@ -90,74 +94,294 @@ pub fn packlist(pkg_dir: &Path, manifest: &Value) -> Result<Vec<String>, Packlis
...
+        }
+        let bundle_manifest = safe_read_package_json_from_dir(&bundle_pkg_dir)
+            .ok()
+            .flatten()
+            .unwrap_or_else(|| Value::Object(serde_json::Map::new()));
+        let nested = packlist(&bundle_pkg_dir, &bundle_manifest)?;
+        for rel in nested {
+            out.insert(format!("node_modules/{bundle_name}/{rel}"));
+        }
+    }
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/git-fetcher/src/packlist.rs:278]: Fixed in 141b8cde. Refactored `packlist` into a public wrapper + private `packlist_inner` that threads a `HashSet<PathBuf>` of canonical visited paths plus a `depth: u32` counter through every recursion. `fs::canonicalize` resolves symlinks so a loop shows up as the same path; `MAX_BUNDLE_DEPTH = 32` is a belt-and-braces guard for any cycle the canonical-path check can't see. Test: `bundle_dependencies_self_cycle_is_caught` plants a symlink loop through `node_modules/self/node_modules/self`, declares a self-bundling manifest, asserts the recursion is cut after one level.

File: crates/git-fetcher/src/packlist.rs [hunk 6]
```diff
@@ -90,74 +94,294 @@ pub fn packlist(pkg_dir: &Path, manifest: &Value) -> Result<Vec<String>, Packlis
...
+    // overridden in pass 1.)
+    if let Some(main) = main_path {
+        let m = normalize_field_path(main);
+        if !m.is_empty() && pkg_dir.join(&m).is_file() {
+            out.insert(m);
+        }
+    }
+    for bin in &bin_paths {
+        let b = normalize_field_path(bin);
+        if !b.is_empty() && pkg_dir.join(&b).is_file() {
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/git-fetcher/src/packlist.rs]: Fixed in 141b8cde. Both `main` and `bin` paths now run through `should_always_exclude` before insertion in pass 3, so an always-excluded basename / VCS segment wins over the manifest field. Tests: `main_field_pointing_at_always_excluded_basename_is_refused` covers the lockfile case (`"main": "package-lock.json"`) and `bin_field_pointing_at_vcs_segment_is_refused` covers VCS-segment paths (`"bin": {"weird": ".git/hook"}`). Matches npm-packlist's silent override.
